packages/core/src/actions/contracts/prepareWriteContract.ts

```diff
@@ -64,9 +65,21 @@ export async function prepareWriteContract<
   overrides,
   signer: signer_,
 }: PrepareWriteContractConfig): Promise<PrepareWriteContractResult<TSigner>> {
-  const signer = signer_ ?? (await fetchSigner())
+  const signer = signer_ ?? (await fetchSigner({ chainId }))
   if (!signer) throw new ConnectorNotFoundError()
 
+  const { chain: activeChain, chains } = getNetwork()
+  const activeChainId = activeChain?.id
+  if (chainId && chainId !== activeChainId) {
+    throw new ChainMismatchError({
+      activeChain:
+        chains.find((x) => x.id === activeChainId)?.name ??
+        `Chain ${activeChainId}`,
+      targetChain:
+        chains.find((x) => x.id === chainId)?.name ?? `Chain ${chainId}`,
+    })
+  }
```

Comment:
Since we have this logic in a multiple places, I wonder if it makes sense to extract it out into an `assertActiveChain` utility?

Reply:
Was thinking the same. Can do!